text/0033-crash-recovery.md

```diff
@@ -0,0 +1,108 @@
+
+This CEP aims to propose a solution that would allow the validators to start voting in eras that are already in progress, making it possible for the network to resume normal operation after temporary issues.
+
+**Note 2020-01-11:** We currently have a quick workaround in place, which basically does what is described below, except it doesn't cater to endorsements, nor does it use storage - it bypasses the storage component by directly creating additional files. While working in practice for the moment, it should be replaced by a proper solution.
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Since the main problem is the risk of creating an accidental equivocation, we need to make sure that a rejoining validator will be aware of all its previous messages in the relevant era.
+
+For the units, it is enough to save the hash of the latest created unit. This would make it possible to download the unit from other validators along with its whole history and be able to verify the correctness of downloaded data.
+
+Units aren't the only thing that might be used to create an equivocation in Highway, though. Endorsements are such a thing, too. It would therefore be necessary to save the hashes of all endorsed units, too. When rejoining the era, the validator would read the list of such hashes, download the relevant units and make sure that it never endorses conflicting units in the future.
+
+The only remaining detail is how to implement saving the relevant hashes in practice. The simplest solution of just saving the hash whenever we create a unit or endorsement has a significant drawback. Imagine we create a unit, save its hash as the hash of the last unit created by us, and then crash before sending the unit to any validator (for any reason, eg. a power outage). In such a case, after restarting, we wouldn't be able to download it again. We wouldn't be able to make sure that we don't equivocate and so we wouldn't start voting again. It is probably the most secure solution, though. An alternative will be discussed in [Rationale and alternatives](#rationale-and-alternatives).
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The consensus component should:
+
+- Whenever a unit or an endorsement is created, emit a request to storage to store the latest data for the era. The data stored would contain:
+    - The hash of the latest unit
+    - The list of hashes of endorsed units
+
+This data should be keyed by the consensus instance ID.
+
+In order to facilitate removing data for old eras (ones that are older than `UNBONDING_PERIOD`), we should also store a map of era ID → instance ID - as instance IDs themselves carry no information about how old the era is. We can, however, identify the era IDs of old eras, and using this map, we could also find the entries to be removed.
```

Comment:
Maybe an easier approach would be to put an expiration date into the file itself.
That sounds dangerous because we only know an _earliest_ date of when the validators will be unbonded, but not a latest one.

I think the solution is to define some constant, say "2 weeks", and:
* When creating a file for an era, put the expiration date into it, which is the earliest day where the era could expire, plus 2 weeks.
* Periodically delete all files with expiration date in the past.
* If we actually happen to instantiate an era whose expiration date would already be in the past, don't participate in it at all. (I.e. always assume we _did_ have a file and deleted it.)

Reply:
I would prefer not to rely on specific timestamps where we can't be certain that they will make sense, especially when we have other options. This approach might use less state than what I proposed, but I think it might turn out to be more complex in the end.